tortoise/verifying_tortoise.go

```diff
@@ -194,428 +309,936 @@ func (t *turtle) checkBlockAndGetInputVector(
 		if err != nil {
-			t.logger.With().Debug("input vector is empty, adding neutral diffs", i, log.Err(err))
+			// an error here signifies a real database failure
+			logger.With().Error("unable to calculate local opinion for layer", log.Err(err))
+			return nil, err
+		}
+
+		// otherwise, nil means we should abstain
+		if layerInputVector == nil {
+			// still waiting for Hare results, vote neutral and move on
+			logger.With().Debug("input vector is empty, adding neutral diffs", log.Err(err))
 			for _, b := range layerBlockIds {
-				if v, ok := opinion2.BlocksOpinion[b]; !ok || v != abstain {
-					t.logger.With().Debug("added diff",
-						log.FieldNamed("base_block_candidate", blockid),
-						log.FieldNamed("diff_block", b),
-						log.String("diff", "neutral"))
-					neutralDiff[b] = struct{}{}
-				}
+				addDiffs(b, "neutral", abstain, neutralDiff)
 			}
 			continue
 		}
+		logger.With().Debug("got local opinion vector for layer", log.Int("count", len(layerInputVector)))
 
 		inInputVector := make(map[types.BlockID]struct{})
 
+		// Add diffs FOR blocks that are in the input vector, but where the base block has no opinion or does not
+		// explicitly support the block
 		for _, b := range layerInputVector {
 			inInputVector[b] = struct{}{}
-			if v, ok := opinion2.BlocksOpinion[b]; !ok || v != support {
-				// Block is in input vector but no base block vote or base block doesn't support it:
-				// add diff FOR this block
-				t.logger.With().Debug("added diff",
-					log.FieldNamed("base_block_candidate", blockid),
-					log.FieldNamed("diff_block", b),
-					log.String("diff", "support"))
-				forDiff[b] = struct{}{}
-			}
+			addDiffs(b, "support", support, forDiff)
 		}
 
-		for _, b := range layerBlockIds {
-			// Matches input vector, no diff (i.e., both support)
-			if _, ok := inInputVector[b]; ok {
-				continue
-			}
-
-			// TODO: maybe we don't need this if it is not included
-			if v, ok := opinion2.BlocksOpinion[b]; !ok || v != against {
-				// Layer block has no base block vote or base block supports, but not in input vector:
-				// add diff AGAINST this block
-				t.logger.With().Debug("added diff",
-					log.FieldNamed("base_block_candidate", blockid),
-					log.FieldNamed("diff_block", b),
-					log.String("diff", "against"))
-				againstDiff[b] = struct{}{}
-			}
-		}
+		// Finally, we need to consider the case where the base block supports a block in this layer that is not in our
+		// input vector (e.g., one we haven't seen), by adding a diff against the block.
+		// We do not explicitly add votes against blocks that the base block does _not_ support, since by not voting to
+		// support a block in an old layer, we are implicitly voting against it. But if the base block does explicitly
+		// support a block and we disagree, we need to add a vote against here.
+		// TODO: this is not currently possible since base block opinions aren't indexed by layer. See
```

Comment:
you can always get full block from db and check the layer though it's quite a lot of overhead.

Reply:
yeah, I'm afraid that would be too expensive. this is a high priority followup item and will be addressed in #2424.